src/controllers/authsController.ts

```diff
@@ -0,0 +1,200 @@
+
+    if(companyExist){
+      return res.status(200).send("Company already exist");
+    } else { 
+
+      await newCompany.save();
+      res.status(200).send("Company created successfully")
+    }
+
+  } catch (err) {
+    next(err);
+  }
+}
+
+
+export const updateCompanyPassword = async(req: Request, res: Response, next: NextFunction) => {
+  
+  if(req.cookies.cookies.id === req.params.companyId){
+
+    try {
+      const salt = bcrypt.genSaltSync(10);
+      const hash = bcrypt.hashSync(req.body.password, salt);
+  
+      await Company.findByIdAndUpdate(
+        req.params.companyId, {$set: {password: hash}}, {new: true} 
+      )
+      res.status(200).send("Password updated successfully");
+    } catch (err) {
+      next(err)
+    }
+  }else {
+    res.status(401).send("You are not authorised to perform this action")
+  }
+}
+
+export const registerOfficer = async(req: Request, res: Response, next: NextFunction) => {
+  const { name, address, companyId, location, phoneNumber, password}: RegisterOfficerType = req.body;
+   
+  if(req.cookies.cookies.id === companyId && req.cookies.cookies.isAdmin){
+
+    try {
+      const salt = bcrypt.genSaltSync(10);
+      const hash = bcrypt.hashSync(password, salt);
+      
+      const newOfficer = new Officer<RegisterOfficerType>({
+        name,
+        address, 
+        companyId,
+        location,
+        phoneNumber,
+        password: hash,
+      })
+      
+      const officerExist: OfficerType | null = await Officer.findOne({phoneNumber});
+      
+      if(officerExist){
+        
+        return res.status(200).send("Officer with the phone already exist")
+      } else {
+        await newOfficer.save();
+        
+        await Company.findByIdAndUpdate(companyId, { $push: {offices: newOfficer}})
+        const { password, ...officerDetail } = newOfficer
+        res.status(200).send("Officer created successfully");
+      }
+    } catch (err) {
+      next(err)
+    }    
+  }else{
+    res.status(401).send("You are not authorised to perform this action")
+  }
+}
+      
+export const updateOfficerPassword = async(req: Request, res: Response, next: NextFunction) => {
+  
+  if(req.cookies.cookies.id === req.params.officerId){
+
+    try {
+      const salt = bcrypt.genSaltSync(10);
+      const hash = bcrypt.hashSync(req.body.password, salt);
+  
+       await Officer.findByIdAndUpdate(
+         req.params.officerId, {$set: {password: hash}}, {new: true} 
+         )
+      res.status(200).send("Password updated successfully");
+    } catch (err) {
+      next(err)
+    }
+  }else {
+    res.status(401).send("You are not authorised to perform this action")
+  }
+}
+  
+export const companyLogin = async(req: Request, res: Response, next: NextFunction) => {  
+    
+  const company: CompanyType | null = await Company.findOne({email: req.body.email});
+  const officer: OfficerType | null = await Officer.findOne({phoneNumber: req.body.phoneNumber});
```

Comment:
```suggestion
  const [company, officer] = await Promise.all([
    Company.findOne({email: req.body.email}),
    Officer.findOne({phoneNumber: req.body.phoneNumber}),
  ]);
```

Reply:
> I will need your help here

 Does this help?

Reply:
very well. it helped.

I have also added caching using Redis to the application